File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupContactsHandler.java
package closer.vlllage.com.closer.handler.group;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;

import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.api.models.PhoneResult;
import closer.vlllage.com.closer.api.models.SuccessResult;
import closer.vlllage.com.closer.handler.data.AccountHandler;
import closer.vlllage.com.closer.handler.data.ApiHandler;
import closer.vlllage.com.closer.handler.data.LocationHandler;
import closer.vlllage.com.closer.handler.data.PermissionHandler;
import closer.vlllage.com.closer.handler.data.PersistenceHandler;
import closer.vlllage.com.closer.handler.data.PhoneContactsHandler;
import closer.vlllage.com.closer.handler.data.RefreshHandler;
import closer.vlllage.com.closer.handler.helpers.ActivityHandler;
import closer.vlllage.com.closer.handler.helpers.AlertHandler;
import closer.vlllage.com.closer.handler.helpers.DefaultAlerts;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.MenuHandler;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.helpers.TimeAgo;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.handler.map.SetNameHandler;
import closer.vlllage.com.closer.handler.phone.NameHandler;
import closer.vlllage.com.closer.handler.phone.PhoneMessagesHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.GroupContact;
import closer.vlllage.com.closer.store.models.GroupContact_;
import closer.vlllage.com.closer.store.models.GroupInvite;
import closer.vlllage.com.closer.store.models.GroupInvite_;
import closer.vlllage.com.closer.store.models.Phone;
import closer.vlllage.com.closer.store.models.Phone_;
import io.objectbox.android.AndroidScheduler;
import io.objectbox.reactive.DataSubscription;
import io.reactivex.Single;

import static android.Manifest.permission.READ_CONTACTS;

public class GroupContactsHandler extends PoolMember {

    private RecyclerView contactsRecyclerView;
    private View showPhoneContactsButton;
    private EditText searchContacts;
    private PhoneContactAdapter phoneContactAdapter;
    private final Set<String> currentGroupContacts = new HashSet<>();
    private DataSubscription dataSubscription;

    @SuppressWarnings("MissingPermission")
    public void attach(Group group, RecyclerView contactsRecyclerView, EditText searchContacts, View showPhoneContactsButton) {
        this.contactsRecyclerView = contactsRecyclerView;
        this.showPhoneContactsButton = showPhoneContactsButton;
        this.searchContacts = searchContacts;
        phoneContactAdapter = new PhoneContactAdapter(this, phoneContact -> {
            if (phoneContact.getName() == null) {
                $(AlertHandler.class).make()
                        .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.invite))
                        .setLayoutResId(R.layout.invite_by_number_modal)
                        .setTextView(R.id.input, name -> {
                            phoneContact.setName(name);
                            inviteToGroup(group, phoneContact);
                        })
                        .setTitle($(ResourcesHandler.class).getResources().getString(R.string.invite_to_group, group.getName()))
                        .show();
            } else {
                $(AlertHandler.class).make()
                        .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.add_phone_name, phoneContact.getFirstName()))
                        .setMessage(phoneContact.getPhoneNumber())
                        .setPositiveButtonCallback(alertResult -> inviteToGroup(group, phoneContact))
                        .setTitle($(ResourcesHandler.class).getResources().getString(R.string.add_phone_to_group, phoneContact.getFirstName(), group.getName()))
                        .show();
            }
        }, groupInvite -> {
            $(AlertHandler.class).make()
                    .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.cancel_invite))
                    .setMessage($(ResourcesHandler.class).getResources().getString(R.string.confirm_cancel_invite, groupInvite.getName()))
                    .setPositiveButtonCallback(alertResult -> {
                        cancelInvite(groupInvite);
                    })
                    .show();
        }, groupContact -> {
            if ($(PersistenceHandler.class).getPhoneId().equals(groupContact.getContactId())) {
                $(MenuHandler.class).show(new MenuHandler.MenuOption(R.drawable.ic_close_black_24dp, R.string.leave_group_action, () -> {
                    $(AlertHandler.class).make()
                            .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.leave_group, group.getName()))
                            .setPositiveButtonCallback(result -> leaveGroup(group))
                            .setTitle($(ResourcesHandler.class).getResources().getString(R.string.leave_group_title, group.getName()))
                            .setMessage($(ResourcesHandler.class).getResources().getString(R.string.leave_group_message))
                            .show();
                }));
            } else {
                $(PhoneMessagesHandler.class).openMessagesWithPhone(groupContact.getContactId(), groupContact.getContactName(), "");
            }
        });

        if($(PermissionHandler.class).has(READ_CONTACTS)) {
            $(DisposableHandler.class).add(
                    $(PhoneContactsHandler.class).getAllContacts().subscribe(phoneContactAdapter::setContacts)
            );
        }

        contactsRecyclerView.setAdapter(phoneContactAdapter);
        contactsRecyclerView.setLayoutManager(new LinearLayoutManager(
                contactsRecyclerView.getContext(),
                LinearLayoutManager.VERTICAL,
                false
        ));

        searchContacts.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
                showContactsForQuery(charSequence.toString());
            }

            @Override
            public void afterTextChanged(Editable editable) {

            }
        });

        $(DisposableHandler.class).add($(StoreHandler.class).getStore().box(GroupInvite.class).query()
                .equal(GroupInvite_.group, group.getId())
                .build().subscribe().on(AndroidScheduler.mainThread()).observer(groupInvites -> {
                    phoneContactAdapter.setInvites(groupInvites);
                }));

        $(DisposableHandler.class).add($(StoreHandler.class).getStore().box(GroupContact.class).query()
                .equal(GroupContact_.groupId, group.getId())
                .build().subscribe().on(AndroidScheduler.mainThread()).observer(groupContacts -> {
                    phoneContactAdapter.setGroupContacts(groupContacts);
                }));
    }

    private void leaveGroup(Group group) {
        $(DisposableHandler.class).add($(ApiHandler.class).leaveGroup(group.getId()).subscribe(successResult -> {
            if (successResult.success) {
                $(DefaultAlerts.class).message(
                        $(ResourcesHandler.class).getResources().getString(R.string.group_no_more, group.getName()),
                        ignored -> $(ActivityHandler.class).getActivity().finish());
                $(RefreshHandler.class).refreshMyGroups();
            } else {
                $(DefaultAlerts.class).thatDidntWork();
            }
        }, error -> $(DefaultAlerts.class).thatDidntWork()));
    }

    private void cancelInvite(GroupInvite groupInvite) {
        $(DisposableHandler.class).add($(ApiHandler.class).cancelInvite(groupInvite.getGroup(), groupInvite.getId()).subscribe(successResult -> {
            if (successResult.success) {
                $(AlertHandler.class).make()
                        .setMessage($(ResourcesHandler.class).getResources().getString(R.string.invite_cancelled))
                        .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.ok))
                        .show();
                $(RefreshHandler.class).refreshMyGroups();
            } else {
                $(DefaultAlerts.class).thatDidntWork(successResult.error);
            }
        }));
    }

    private void inviteToGroup(Group group, PhoneContact phoneContact) {
        String myName = $(AccountHandler.class).getName();
        if (myName == null || myName.trim().isEmpty()) {
            $(SetNameHandler.class).modifyName(name -> sendInviteToGroup(group, phoneContact), true);
        } else {
            sendInviteToGroup(group, phoneContact);
        }
    }

    private void sendInviteToGroup(Group group, PhoneContact phoneContact) {
        Single<SuccessResult> inviteToGroup = phoneContact.getPhoneId() == null ?
                $(ApiHandler.class).inviteToGroup(group.getId(), phoneContact.getName(), phoneContact.getPhoneNumber()) :
                $(ApiHandler.class).inviteToGroup(group.getId(), phoneContact.getPhoneId());

        $(DisposableHandler.class).add(inviteToGroup.subscribe(successResult -> {
                    if (successResult.success) {
                        String message = phoneContact.getName() == null || phoneContact.getName().trim().isEmpty() ?
                                $(ResourcesHandler.class).getResources().getString(R.string.phone_invited, phoneContact.getPhoneNumber(), group.getName()) :
                                $(ResourcesHandler.class).getResources().getString(R.string.phone_invited, phoneContact.getName(), group.getName());
                        $(AlertHandler.class).make()
                                .setMessage(message)
                                .setPositiveButton($(ResourcesHandler.class).getResources().getString(R.string.yaay))
                                .show();
                        $(RefreshHandler.class).refreshMyGroups();
                    } else {
                        $(DefaultAlerts.class).thatDidntWork(successResult.error);
                    }
                }, error -> $(DefaultAlerts.class).thatDidntWork()));
    }

    public boolean isEmpty() {
        return phoneContactAdapter.getItemCount() == 0;
    }

    @SuppressWarnings("MissingPermission")
    public void showContactsForQuery(String originalQuery) {
        String query = originalQuery.trim().toLowerCase();

        if ($(LocationHandler.class).getLastKnownLocation() != null) {
            $(DisposableHandler.class).add($(ApiHandler.class).searchPhonesNear(new LatLng(
                    $(LocationHandler.class).getLastKnownLocation().getLatitude(),
                    $(LocationHandler.class).getLastKnownLocation().getLongitude()
            ), query).subscribe(phoneResults -> {
                    for (PhoneResult phoneResult : phoneResults) {
                        $(RefreshHandler.class).refresh(PhoneResult.from(phoneResult));
                    }
            }, error -> $(DefaultAlerts.class).thatDidntWork()));
        }

        String phoneNumber = $(PhoneNumberHandler.class).normalize(originalQuery);

        phoneContactAdapter.setPhoneNumber(phoneNumber);
        phoneContactAdapter.setIsFiltered(!originalQuery.isEmpty());

        if(!$(PermissionHandler.class).has(READ_CONTACTS)) {
            showPhoneContacts(new ArrayList<>(), query);
            return;
        }

        $(DisposableHandler.class).add($(PhoneContactsHandler.class).getAllContacts().subscribe(phoneContacts -> {
            showPhoneContacts(phoneContacts, query);
        }, error -> $(DefaultAlerts.class).thatDidntWork()));
    }

    private void showPhoneContacts(List<PhoneContact> phoneContacts, String query) {
        if (dataSubscription != null) {
            $(DisposableHandler.class).dispose(dataSubscription);
        }
        dataSubscription = $(StoreHandler.class).getStore().box(Phone.class).query()
                .contains(Phone_.name, $(Val.class).of(query))
                .notNull(Phone_.id)
                .greater(Phone_.updated, $(TimeAgo.class).fifteenDaysAgo())
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer(closerContacts -> {
                    List<PhoneContact> allContacts = new ArrayList<>();

                    if (phoneContacts != null) {
                        allContacts.addAll(phoneContacts);
                    }

                    for (Phone phone : closerContacts) {
                        if (currentGroupContacts.contains(phone.getId())) {
                            continue;
                        }

                        allContacts.add(0, new PhoneContact($(NameHandler.class).getName(phone), phone.getStatus()).setPhoneId(phone.getId()));
                    }

                    if (query.isEmpty()) {
                        phoneContactAdapter.setContacts(allContacts);
                        return;
                    }

                    String queryPhone = query.replaceAll("[^0-9]", "");

                    List<PhoneContact> contacts = new ArrayList<>();
                    for(PhoneContact contact : allContacts) {
                        if (contact.getName() != null) {
                            if (contact.getName().toLowerCase().contains(query)) {
                                contacts.add(contact);
                                continue;
                            }
                        }

                        if (!queryPhone.isEmpty() && contact.getPhoneNumber() != null) {
                            if (contact.getPhoneNumber().replaceAll("[^0-9]", "").contains(queryPhone)) {
                                contacts.add(contact);
                            }
                        }
                    }

                    phoneContactAdapter.setContacts(contacts);
                });
        $(DisposableHandler.class).add(dataSubscription);
    }

    public void showContactsForQuery() {
        showContactsForQuery(searchContacts.getText().toString());
    }

    public void setCurrentGroupContacts(List<GroupContact> groupContacts) {
        currentGroupContacts.clear();
        for (GroupContact groupContact : groupContacts) {
            currentGroupContacts.add(groupContact.getContactId());
        }
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupContactsHandler.kt
package closer.vlllage.com.closer.handler.group

import android.Manifest.permission.READ_CONTACTS
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.data.*
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.handler.map.SetNameHandler
import closer.vlllage.com.closer.handler.phone.NameHandler
import closer.vlllage.com.closer.handler.phone.NavigationHandler
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.*
import at.bluesource.choicesdk.maps.common.LatLng
import closer.vlllage.com.closer.databinding.InviteByNumberModalBinding
import com.queatz.on.On
import io.objectbox.android.AndroidScheduler
import io.objectbox.query.QueryBuilder
import io.objectbox.reactive.DataSubscription
import java.util.*

class GroupContactsHandler constructor(private val on: On) {

    private lateinit var contactsRecyclerView: RecyclerView
    private lateinit var showPhoneContactsButton: View
    private lateinit var searchContacts: EditText
    private lateinit var phoneContactAdapter: PhoneContactAdapter
    private val currentGroupContacts = HashSet<String>()
    private var dataSubscription: DataSubscription? = null

    val isEmpty: Boolean get() = phoneContactAdapter.itemCount == 0

    fun attach(group: Group, contactsRecyclerView: RecyclerView, searchContacts: EditText, showPhoneContactsButton: View) {
        this.contactsRecyclerView = contactsRecyclerView
        this.showPhoneContactsButton = showPhoneContactsButton
        this.searchContacts = searchContacts
        phoneContactAdapter = PhoneContactAdapter(on, { phoneContact ->
            if (phoneContact.name == null) {
                on<AlertHandler>().view { InviteByNumberModalBinding.inflate(it) }.apply {
                    positiveButton = on<ResourcesHandler>().resources.getString(R.string.invite)
                    textViewId = R.id.input
                    onTextViewSubmitCallback = { name ->
                        phoneContact.name = name
                        inviteToGroup(group, phoneContact)
                    }
                    title = on<ResourcesHandler>().resources.getString(R.string.invite_to_group, group.name)
                    show()
                }
            } else {
                on<AlertHandler>().make().apply {
                    positiveButton = on<ResourcesHandler>().resources.getString(R.string.add_phone_name, phoneContact.firstName)
                    message = phoneContact.phoneNumber
                    positiveButtonCallback = { inviteToGroup(group, phoneContact) }
                    if (phoneContact.phoneId != null) {
                        negativeButton = on<ResourcesHandler>().resources.getString(R.string.profile)
                        negativeButtonCallback = { on<NavigationHandler>().showProfile(phoneContact.phoneId!!) }
                    }
                    title = on<ResourcesHandler>().resources.getString(R.string.add_phone_to_group, phoneContact.firstName, group.name)
                    show()
                }
            }
        }, { groupInvite ->
            on<AlertHandler>().make().apply {
                positiveButton = on<ResourcesHandler>().resources.getString(R.string.cancel_invite)
                message = on<ResourcesHandler>().resources.getString(R.string.confirm_cancel_invite, groupInvite.name)
                positiveButtonCallback = { cancelInvite(groupInvite) }
                show()
            }
        }, { groupContact ->
            if (on<PersistenceHandler>().phoneId == groupContact.contactId) {
                on<MenuHandler>().show(
                        MenuHandler.MenuOption(R.drawable.ic_person_black_24dp, R.string.view_profile) {
                            on<NavigationHandler>().showMyProfile()
                        },
                        MenuHandler.MenuOption(R.drawable.ic_chat_black_24dp, R.string.change_your_group_status) {
                            on<DefaultInput>().show(
                                    R.string.your_group_status,
                                    buttonRes = R.string.update,
                                    prefill = groupContact.status
                            ) {
                                on<GroupContactHandler>().updateGroupStatus(groupContact, it)
                            }
                        },
                        MenuHandler.MenuOption(R.drawable.ic_photo_black_24dp, R.string.change_your_group_photo) {
                            on<DefaultMenus>().uploadPhoto { photoId ->
                                val photo = on<PhotoUploadGroupMessageHandler>().getPhotoPathFromId(photoId)
                                on<GroupContactHandler>().updateGroupPhoto(groupContact, photo)
                            }
                        },
                        MenuHandler.MenuOption(R.drawable.ic_close_black_24dp, R.string.leave_group_action) {
                    on<AlertHandler>().make().apply {
                        negativeButton = on<ResourcesHandler>().resources.getString(R.string.nope)
                        positiveButton = on<ResourcesHandler>().resources.getString(R.string.leave_group, group.name)
                        positiveButtonCallback = { leaveGroup(group) }
                        title = on<ResourcesHandler>().resources.getString(R.string.leave_group_title, group.name)
                        message = on<ResourcesHandler>().resources.getString(
                                if (group.isPublic) R.string.leave_public_group_message else R.string.leave_private_group_message)
                        show()
                    }
                })
            } else {
                on<NavigationHandler>().showProfile(groupContact.contactId!!)
            }
        })

        if (on<PermissionHandler>().has(READ_CONTACTS)) {
            on<DisposableHandler>().add(
                    on<PhoneContactsHandler>().allContacts.subscribe { phoneContactAdapter.setContacts(it) }
            )
        }

        contactsRecyclerView.adapter = phoneContactAdapter
        contactsRecyclerView.layoutManager = LinearLayoutManager(
                contactsRecyclerView.context,
                RecyclerView.VERTICAL,
                false
        )

        searchContacts.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {

            }

            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                showContactsForQuery(charSequence.toString())
            }

            override fun afterTextChanged(editable: Editable) {

            }
        })

        on<DisposableHandler>().add(on<StoreHandler>().store.box(GroupInvite::class).query()
                .equal(GroupInvite_.group, group.id!!, QueryBuilder.StringOrder.CASE_SENSITIVE)
                .build().subscribe().on(AndroidScheduler.mainThread()).observer { groupInvites -> phoneContactAdapter.setInvites(groupInvites) })

        on<DisposableHandler>().add(on<StoreHandler>().store.box(GroupContact::class).query()
                .equal(GroupContact_.groupId, group.id!!, QueryBuilder.StringOrder.CASE_SENSITIVE)
                .build().subscribe().on(AndroidScheduler.mainThread()).observer { groupContacts -> phoneContactAdapter.setGroupContacts(groupContacts) })
    }

    fun leaveGroup(group: Group) {
        on<DisposableHandler>().add(on<ApiHandler>().leaveGroup(group.id!!).subscribe({ successResult ->
            if (successResult.success) {
                on<DefaultAlerts>().message(
                        on<ResourcesHandler>().resources.getString(R.string.group_no_more, group.name)
                ) { on<ActivityHandler>().activity!!.finish() }
                on<RefreshHandler>().refreshMyGroups()
            } else {
                on<DefaultAlerts>().thatDidntWork()
            }
        }, { error -> on<DefaultAlerts>().thatDidntWork() }))
    }

    private fun cancelInvite(groupInvite: GroupInvite) {
        on<DisposableHandler>().add(on<ApiHandler>().cancelInvite(groupInvite.group!!, groupInvite.id!!).subscribe({ successResult ->
            if (successResult.success) {
                on<AlertHandler>().make().apply {
                    message = on<ResourcesHandler>().resources.getString(R.string.invite_cancelled)
                    positiveButton = on<ResourcesHandler>().resources.getString(R.string.ok)
                    show()
                }
                on<RefreshHandler>().refreshMyGroups()
            } else {
                on<DefaultAlerts>().thatDidntWork(successResult.error)
            }
        }) { on<DefaultAlerts>().thatDidntWork((it as? ApiError)?.error) })
    }

    private fun inviteToGroup(group: Group, phoneContact: PhoneContact) {
        val myName = on<AccountHandler>().name
        if (myName.isBlank()) {
            on<SetNameHandler>().modifyName({
                sendInviteToGroup(group, phoneContact)
            }, true)
        } else {
            sendInviteToGroup(group, phoneContact)
        }
    }

    private fun sendInviteToGroup(group: Group, phoneContact: PhoneContact) {
        val inviteToGroup = if (phoneContact.phoneId == null)
            on<ApiHandler>().inviteToGroup(group.id!!, phoneContact.name!!, phoneContact.phoneNumber!!)
        else
            on<ApiHandler>().inviteToGroup(group.id!!, phoneContact.phoneId!!)

        on<DisposableHandler>().add(inviteToGroup.subscribe({ successResult ->
            if (successResult.success) {
                val message = if (phoneContact.name == null || phoneContact.name!!.isBlank())
                    on<ResourcesHandler>().resources.getString(R.string.phone_invited, phoneContact.phoneNumber, group.name)
                else
                    on<ResourcesHandler>().resources.getString(R.string.phone_invited, phoneContact.name, group.name)
                on<ToastHandler>().show(message)
                on<RefreshHandler>().refreshMyGroups()
            } else {
                on<DefaultAlerts>().thatDidntWork(successResult.error)
            }
        }, { on<DefaultAlerts>().thatDidntWork((it as? ApiError)?.error) }))
    }

    fun showContactsForQuery(originalQuery: String) {
        val query = originalQuery.trim().toLowerCase()

        if (on<LocationHandler>().lastKnownLocation != null) {
            on<DisposableHandler>().add(on<ApiHandler>().searchPhonesNear(LatLng(
                    on<LocationHandler>().lastKnownLocation!!.latitude,
                    on<LocationHandler>().lastKnownLocation!!.longitude
            ), query).subscribe({ phoneResults ->
                for (phoneResult in phoneResults) {
                    on<RefreshHandler>().refresh(on<ApiModelHandler>().from(phoneResult))
                }
            }, { error -> on<DefaultAlerts>().thatDidntWork() }))
        }

        val phoneNumber = on<PhoneNumberHandler>().normalize(originalQuery)

        phoneContactAdapter.setPhoneNumber(phoneNumber)
        phoneContactAdapter.setIsFiltered(!originalQuery.isEmpty())

        if (!on<PermissionHandler>().has(READ_CONTACTS)) {
            showPhoneContacts(listOf(), query)
            return
        }

        on<DisposableHandler>().add(on<PhoneContactsHandler>().allContacts.subscribe({ phoneContacts -> showPhoneContacts(phoneContacts, query) }, { error -> on<DefaultAlerts>().thatDidntWork() }))
    }

    private fun showPhoneContacts(phoneContacts: List<PhoneContact>?, query: String) {
        if (dataSubscription != null) {
            on<DisposableHandler>().dispose(dataSubscription!!)
        }
        dataSubscription = on<StoreHandler>().store.box(Phone::class).query()
                .contains(Phone_.name, on<Val>().trimmed(query), QueryBuilder.StringOrder.CASE_SENSITIVE)
                .notNull(Phone_.id)
                .greater(Phone_.updated, on<TimeAgo>().fifteenDaysAgo())
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer { closerContacts ->
                    val allContacts = mutableListOf<PhoneContact>()

                    if (phoneContacts != null) {
                        allContacts.addAll(phoneContacts)
                    }

                    for (phone in closerContacts) {
                        if (currentGroupContacts.contains(phone.id)) {
                            continue
                        }

                        allContacts.add(0, PhoneContact(on<NameHandler>().getName(phone), phone.status).apply {
                            phoneId = phone.id
                        })
                    }

                    if (query.isEmpty()) {
                        phoneContactAdapter.setContacts(allContacts)
                        return@observer
                    }

                    val queryPhone = query.replace("[^0-9]".toRegex(), "")

                    val contacts = mutableListOf<PhoneContact>()
                    for (contact in allContacts) {
                        if (contact.name != null) {
                            if (contact.name!!.toLowerCase().contains(query)) {
                                contacts.add(contact)
                                continue
                            }
                        }

                        if (queryPhone.isNotEmpty() && contact.phoneNumber != null) {
                            if (contact.phoneNumber!!.replace("[^0-9]".toRegex(), "").contains(queryPhone)) {
                                contacts.add(contact)
                            }
                        }
                    }

                    phoneContactAdapter.setContacts(contacts)
                }
        on<DisposableHandler>().add(dataSubscription!!)
    }

    fun showContactsForQuery() {
        showContactsForQuery(searchContacts.text.toString())
    }

    fun setCurrentGroupContacts(groupContacts: List<GroupContact>) {
        currentGroupContacts.clear()
        for (groupContact in groupContacts) {
            currentGroupContacts.add(groupContact.contactId!!)
        }
    }
}
